### PR TITLE
Docment losses

### PR DESCRIPTION
**Pull request by kevinbird15**
Planning on breaking this up into a few chunks to get some initial feedback.  I would like to just keep this PR open but get a first round of feedback on the losses I've docmented so far.  I did encounter one issue with the HTML rendering so I would call that a known issue, but besides that, just looking for general advice and an accuracy check on what I've put so far.

### CODE REVIEW

File: nbs/01a_losses.ipynb [hunk 1]
```diff
@@ -62,7 +62,15 @@
     "class BaseLoss():\n",
```
Comment (Isaac-Flath) [inline nbs/01a_losses.ipynb:62]: <pre class="ql-syntax" spellcheck="false">Line #8.            axis:int=-1, # The axis on which to reduce
</pre><p>Should docment just say <code># reduction axis</code>  to be more concise? </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/muellerzr/fastai-docment-sprint/pull/87/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (Isaac-Flath) [inline nbs/01a_losses.ipynb:62]: <p>Not docments related, but it would be very helpful to include requirements of the <code>loss_cls</code>  function.  What must that function take as input?  What must it return? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/muellerzr/fastai-docment-sprint/pull/87/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (Isaac-Flath) [inline nbs/01a_losses.ipynb:62]: <pre class="ql-syntax" spellcheck="false">Line #9.            axis:int=-1, #  Determines which axis on which self.decodes and self.activation occurs

</pre><p>Slight shortening <code>#  Axis self.decodes and self.activation occurs on</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/muellerzr/fastai-docment-sprint/pull/87/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (Isaac-Flath) [inline nbs/01a_losses.ipynb:62]: <pre class="ql-syntax" spellcheck="false">Line #35.            axis:int=-1, # Determines which axis on which self.decodes and self.activation occurs
</pre><p> </p><p>Can we shorten docment to <code># Axis self.decodes and self.activation occurs on</code></p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/muellerzr/fastai-docment-sprint/pull/87/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

Comment (Isaac-Flath) [inline nbs/01a_losses.ipynb:62]: <pre class="ql-syntax" spellcheck="false">Line #8.            axis:int=-1, # The demension that has the dimension
</pre><p> </p><p>I don't understand this docment</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/muellerzr/fastai-docment-sprint/pull/87/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

Comment (Isaac-Flath) [inline nbs/01a_losses.ipynb:62]: <pre class="ql-syntax" spellcheck="false">Line #9.            floatify:bool=True, # Whether targ need to be converted to float
</pre><p> </p><p>Maybe shorten to "# coerce <code>targ</code> to float?" to be more concise</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/muellerzr/fastai-docment-sprint/pull/87/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

Comment (Isaac-Flath) [inline nbs/01a_losses.ipynb:62]: <pre class="ql-syntax" spellcheck="false">Line #20.            "Used on predictions in inference"
</pre><p>My opinion would be to modify the doc string to describe what it does as well as what it acts on.  Maybe</p><p> </p><p><code>Converts raw predictions/probabilities to class predictions</code> or something like that?  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/muellerzr/fastai-docment-sprint/pull/87/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

Comment (kevinbird15) [reply nbs/01a_losses.ipynb:62]: Pretty sure that's because it doesn't make sense.  And I didn't even spell it correctly lol thanks for all your great comments.  It is definitely helping me improve

**Comment by kevinbird15** [comment]
closes #6

File: fastai/losses.py [hunk 2]
```diff
@@ -70,55 +104,95 @@ class FocalLossFlat(BaseLoss):
...
-    y_int = True
+    y_int = True # y interpolation
     @use_kwargs_dict(keep=True, weight=None, reduction='mean')
-    def __init__(self, *args, gamma=2.0, axis=-1, **kwargs):
+    def __init__(self,
+        *args,
+        gamma:float=2.0, # Weighting factor to address class imbalance
+        axis:int=-1, # Axis self.decodes and self.activation occurs on
+        **kwargs
+    ):
```
Comment (warner-benjamin) [inline fastai/losses.py:114]: Focal loss' gamma had the weights doc string, not gamma's
```suggestion
    def __init__(self,
        *args,
        gamma:float=2.0, # Focusing parameter. Higher values down-weight easy examples' contribution to loss
        axis:int=-1, # Input class axis
        **kwargs
    ):
```

**Review by warner-benjamin — CHANGES_REQUESTED** [review]
I provided some examples as suggestions for a few losses. Will need to be replicated across the rest of the losses.

* Lots of methods are missing type hints, particularly return type hints
* Argument doc strings could be more concise. Provided some examples.
* Fixed some incorrect doc strings

File: fastai/losses.py [hunk 3]
```diff
@@ -13,47 +13,81 @@
...
+        **kwargs
+    ): super().__init__(nn.CrossEntropyLoss, *args, axis=axis, **kwargs)
+
+    def decodes(self, x):
+        "Converts raw predictions/probabilities to class predictions"
+        return x.argmax(dim=self.axis)
+
+    def activation(self, x):
+        "Last layer's activation"
+        return F.softmax(x, dim=self.axis)
```
[the remark this answers is not in the document]
Comment (kevinbird15) [reply fastai/losses.py:76]: How about something like this?  

```suggestion
    def activation(self, x):
        "Activation function applied on model's output" 
        return F.softmax(x, dim=self.axis)
```

[the remark this answers is not in the document]
Comment (warner-benjamin) [reply fastai/losses.py]: I think it would be nice to concisely mention that the activation function is from the loss function, but otherwise it is an improvement

File: fastai/losses.py (hunk 2 above)
Comment (kevinbird15) [reply fastai/losses.py]: Is `Class axis` appropriate?  I wasn't quite sure what Input referred to so I just did Class axis, but let me know if that changes the intent

File: fastai/losses.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (kevinbird15) [reply fastai/losses.py]: I ended up blending the two.  Let me know if it is acceptable.

File: fastai/losses.py (hunk 2 above)
Comment (warner-benjamin) [reply fastai/losses.py]: Input label axis would probably be the most accurate

File: fastai/losses.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (warner-benjamin) [reply fastai/losses.py]: > `nn.CrossEntropyLoss`'s fused activation function applied to model output

Maybe this? Also, some of the other activations functions didn't get updated

File: fastai/losses.py (hunk 2 above)
Comment (kevinbird15) [reply fastai/losses.py]: I used your suggestion here, but I don't understand why we use the term `input` in this case.

File: fastai/losses.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (kevinbird15) [reply fastai/losses.py]: I didn't see any that hadn't been updated.  I went ahead and changed to the version you suggest here.  Thanks!

File: fastai/losses.py (hunk 2 above)
Comment (warner-benjamin) [reply fastai/losses.py]: On second look, I misread something and you are right, it shouldn't be input

Comment (kevinbird15) [reply fastai/losses.py]: Ok, either `Label axis` or `Class axis` then?

Comment (kevinbird15) [reply fastai/losses.py]: I'm leaning towards `Class axis` but I'll leave it up to you

**Review by warner-benjamin — APPROVED** [review]
LGTM
